🚀 1단계 - 예약 기능 구현(1-2단계 데이터베이스 적용)

junseong94 opened the pull request:
안녕하세요 리뷰어님 

1-2단계 데이터베이스 적용 구현 후 PR 드립니다.

과제 진행중 하기와 같이 질문사항이 있어 짧게 적어봤습니다.

- 개요
Service layer에서 @Transactional을 적용한 메서드 중 delete 함수에서 try-catch문을 적용하였습니다. 
이유는 Database에 해당 예악정보가 있는지 확인 후 데이터가 없다면 exception으로 처리하기 위해 사용했습니다.
하지만 이로 인해 Controller layer에서도 exception 처리가 필요하게 됐고, 
GlobalControllerAdvice(@ControllerAdvice) 클래스에 에러를 반환하도록 했습니다.

- 질문
코드를 작성하면서 Service layer 예외처리를 위해 Controller layer에서도 후속조치를 해줘야하는 과정이 뭔가 불필요한 코드가 늘어난다는 생각을 하게되어 어떤식으로 해결해야할 수 있을지 조언 부탁드립니다

junseong94 committed:
실습 페이지 설정 완료

junseong94 committed:
예약 관리 API - 예약조회 API 구현 완료

junseong94 committed:
예약 관리 API - 예약추가 API 구현 완료

junseong94 committed:
예약 관리 API - 예약취소 API 구현 완료

junseong94 committed:
h2 database 설정 완료

junseong94 committed:
refactor : Reservation, ReservationController
create : ReservationResponseDto
update : ReservationDto to ReservationRequestDto

junseong94 committed:
update : 예악저장소 명세 변경(HashMap to Map)

junseong94 committed:
ReservationRepository : 구현 및 테스트 코드 작성

junseong94 committed:
ReservationService 추가 구현
ReservationController 리팩토링
GlobalExceptionHandler 추가
ReservationController 리팩토링 : HashMap 기반 테스트 코드 일부 수정

Hyeon9mak commented on src/main/java/roomescape/reservation/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,15 @@
+    public ResponseEntity<String> handleException(Exception ex) {
+        return new ResponseEntity<>("An error occurred", HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+}
```
https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/
😉

Hyeon9mak commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,82 @@
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class Reservation {
+    private static final AtomicLong index = new AtomicLong(1L);
+    private Long id;
+    private String name;
+    private String date;
+    private String time;
+
+    public Reservation(String name, String date, String time) {
```
id = 1, name = "혹성탈출" 예약과
id = 1, name = "방탈출" 예약은
서로 같은 예약일까요, 다른 예약일까요?

서로를 같은 예약으로 바라본다면, id 에 대한 동등성 부여가 필요하겠어요!

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    public Long save(Reservation reservation) {
+        final KeyHolder keyHolder = new GeneratedKeyHolder();
+        final String sql = "insert into reservation(name, date, time) values(?,?,?)";
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
+            ps.setString(1, reservation.getName());
+            ps.setString(2, reservation.getDate());
+            ps.setString(3, reservation.getTime());
+            return ps;
+        }, keyHolder);
+
+        return keyHolder.getKey().longValue();
+    }
```
SimpleJdbcInsert 를 활용해보아도 좋을거 같아요!

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,74 @@
+
+
+
+    public Reservation findById(Long id) {
+        final String sql = "select id, name, date, time from reservation where id = ?";
+        final Reservation reservation = jdbcTemplate.queryForObject(sql,
+                (resultSet, rowNum) -> {
+                    Reservation reservation1 = new Reservation(
+                            resultSet.getLong("id"),
+                            resultSet.getString("name"),
+                            resultSet.getString("date"),
+                            resultSet.getString("time")
+                    );
+                    return reservation1;
+                }, id);
+
+        return reservation;
```
매번 Reservation 을 만들어주는 코드가 반복되는데요,
row mapper 를 필드 멤버로 갖도록 하는 것은 어떨까용?

Hyeon9mak commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -0,0 +1,109 @@
+                "PRIMARY KEY (id))");
+    }
+
+    @DisplayName("전체 예약을 조회 합니다.")
+    @Test
+    void readReservation() {
+
+        // given
+        final Reservation reservation1 = new Reservation("제이슨", "2023-08-05", "15:40");
+        final Reservation reservation2 = new Reservation("심슨", "2023-08-05", "15:40");
```
🤣

Hyeon9mak commented on src/main/java/roomescape/reservation/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
```
`@RestControllerAdvice` 와는 어떤 차이를 가질까요?

Hyeon9mak commented on src/main/java/roomescape/reservation/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,15 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(value = Exception.class)
+    public ResponseEntity<String> handleException(Exception ex) {
+        return new ResponseEntity<>("An error occurred", HttpStatus.INTERNAL_SERVER_ERROR);
```
Exception 하위의 다른 예외들 중, 준성님께서 예상하시기에 '이 예외는 발생하겠어.' 라는 것들은 별도로 명세해보는 건 어떨까요?

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+
+    @Transactional
+    public void delete(Long id) {
+        try {
+            final Reservation savedReservaion = reservationRepository.findById(id);
+            reservationRepository.deleteById(savedReservaion.getId());
+        } catch (Exception e) {
+            throw new IllegalArgumentException("해당 예약이 존재하지 않습니다.");
+        }
+    }
```
우선 문제를 해결할 수 있는 방법은 크게 2가지로 생각을 해볼 수 있겠는데요,

### existsById 활용

1. findById 대신 existsById 를 통해 Repository 에게 예약이 있는지 질의하고, boolean 으로 응답을 받는다. (row 가 1 개 이상인지 확인)
2. 참이라면 삭제를 수행하고, 거짓이라면 예외를 던진다.(throw exception)

### deleteById 의 응답 활용

1. 주어진 id 로 곧장 deleteById 를 수행, jdbcTemplate 이 반환하는 정수값을 그대로 service 계층까지 끌고와서, 영향을 받은 row 의 개수가 1개 미만이라면 예외를 던진다. (throw exception)

Hyeon9mak replied:
> 코드를 작성하면서 Service layer 예외처리를 위해 Controller layer에서도 후속조치를 해줘야하는 과정이 뭔가 불필요한 코드가 늘어난다는 생각을 하게되어 어떤식으로 해결해야할 수 있을지 조언 부탁드립니다

spring context 내부에서 발생하는 모든 runtime exception 은 controller advice 를 통해 자동으로 잡아줄 수 있습니다!
때문에 편하게 예외만 throw 하셔도 충분할 것 같아요!

Hyeon9mak reviewed (changes requested):
안녕하세요 준성님! 나머지 구현도 잘 해주셨습니다.
질문 주신 내용 포함하여 코멘트를 남겼으니 확인해주세요!
궁금하신 점 얼마든지 추가 질문 남겨주세요!!

junseong94 committed:
GlobalExceptionHandler : 부가적인 예외처리 코드 추가
Reservation : id 기준 동등성 부여를 위해 equals, hashCode 추가
ReservationRepository : row Mapper 필드로 추가
ReservationService : existsById를 조건식으로 delete 로직 수정

Hyeon9mak commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,96 @@
+        return time;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Reservation that = (Reservation) o;
+        return Objects.equals(id, that.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(id);
+    }
```
😉 👍

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,58 @@
+        );
+    }
+
+    public List<Reservation> findAll() {
+        final String sql = "select id, name, date, time from reservation";
+        return jdbcTemplate.query(sql, reservationRowMapper);
+
+    }
+
+    public Long save(Reservation reservation) {
+        final SqlParameterSource parameters = new BeanPropertySqlParameterSource(reservation);
+        return simpleJdbcInsert.executeAndReturnKey(parameters).longValue();
+    }
+
+    public void deleteById(Long id) {
+        final String sql = "delete from reservation where id = ?";
+        jdbcTemplate.update(sql, Long.valueOf(id));
+    }
```
row mapper 가 멤버변수로 빠진 덕분에 각각 쿼리 수행문이 엄청 간결해졌네요!

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+
+    @Transactional
+    public void delete(Long id) {
+        if(!reservationRepository.existsById(id)) {
+            throw new IllegalArgumentException("해당 예약이 존재하지 않습니다.");
+        } else if (reservationRepository.existsById(id)) {
+            reservationRepository.deleteById(id);
+        }
+    }
```
삭제를 수행할 때 쿼리가 총 3회 수행되겠네요!
질의를 한번 한 후, 삭제가 수행되도록 하여 총 2회만 쿼리가 실행되도록 하면 좋겠어요!

```suggestion
    @Transactional
    public void delete(Long id) {
        final boolean isExistedReservation = reservationRepository.existsById(id);
        if (isExistedReservation) {
            reservationRepository.deleteById(id);
        }
        throw new IllegalArgumentException("해당 예약이 존재하지 않습니다.");
    }
```

Hyeon9mak reviewed (approved):
안녕하세요 준성님! 피드백 잘 반영해주셨습니다 👍 
다음 단계 진행에 참고하실만한 코멘트 남겨드렸으니 확인해주세요!
다음 단계로 넘어가시죠~
